lib/edda-server/src/compressor.rs

```diff
@@ -0,0 +1,422 @@
+        change_set_id: ChangeSetId,
+        workspace_id: WorkspacePk,
+        ctx: DalContext,
+    }
+}
+
+impl Compressor {
+    pub async fn new(
+        requests_stream: &jetstream::stream::Stream,
+        consumer_config: push::OrderedConfig,
+        ctx_builder: DalContextBuilder,
+        change_set_id: ChangeSetId,
+        workspace_id: WorkspacePk,
+    ) -> Result<Self> {
+        let incoming = requests_stream
+            .create_consumer(consumer_config)
+            .await
+            .map_err(CompressorError::ConsumerCreate)?
+            .messages()
+            .await
+            .map_err(CompressorError::Subscribe)?;
+
+        let ctx = ctx_builder
+            .build_for_change_set_as_system(workspace_id, change_set_id, None)
+            .await
+            .map_err(CompressorError::DalContextBuild)?;
+
+        Ok(Self {
+            incoming,
+            state: State::Inactive,
+            ctx_builder,
+            change_set_id,
+            workspace_id,
+            ctx,
+        })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum CompressorRequest {
+    CompressorRebuild(CompressorRebuildRequestVCurrent),
+    CompressorUpdate(CompressorUpdateRequestVCurrent),
+}
+
+impl futures::Stream for Compressor {
+    type Item = std::result::Result<Message, consumer::push::MessagesError>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let project = self.project();
+
+        // FIXME(nick): we need to handle timing or max size. What if we have been pending for a long time or have made a massive batch?
+        // We should publish the compressed message at certain limits. We need to first see if there are too many or too few compression
+        // sets happening first though.
+
+        loop {
+            match project.incoming.poll_next_unpin(cx) {
+                Poll::Ready(Some(result)) => {
+                    // First, get the message off the stream and validate it.
+                    let message = match result {
+                        Ok(message) => message,
+                        Err(err) => {
+                            error!(si.error.message = ?err, "error getting info for jetstream message");
+                            continue;
+                        }
+                    };
+
+                    // FIXME(nick): remove later. Only used to make compiler happy to get this into draft PR form.
+                    let cloned_message = message.clone();
+
+                    // Determine if the stream has been drained.
+                    let drained = match message.info() {
+                        Ok(info) => info.pending == 0,
+                        Err(err) => {
+                            error!(si.error.message = ?err, "error getting info for jetstream message");
+                            continue;
+                        }
+                    };
```

Comment:
Subtle, but this is a trick from @fnichol's recommendation to check the stream info's "pending" count to tell if more things are in the pipe because the `Stream` `impl` may not be sufficient on its own. We could be using this more, but this is a start.